### PR TITLE
Add an abstract interface FaceRestriction

### LINKED ISSUE #2328: Inconsistent behavior of L2FaceRestriction implementation of Operator::MultTranspose
In the documentation for `mfem::Operator::MultTranspose()`: implements "y=A^T(x)" (i.e. direct assignment)

But `L2FaceRestriction::MultTranspose()` appears (?) to be implementing it as "y += A^T(x)" instead (i.e. compound assignment)

```cpp
#include <iostream>

#include "mfem.hpp"

int main(){
  mfem::Mesh mesh = mfem::Mesh(2, 1, mfem::Element::QUADRILATERAL, 1, 2.0, 3.0);

  auto fec = mfem::L2_FECollection(2, 2, mfem::BasisType::GaussLobatto);
  mfem::FiniteElementSpace fespace(&mesh, &fec);

  auto * op = fespace.GetFaceRestriction(mfem::ElementDofOrdering::LEXICOGRAPHIC, 
                                         mfem::FaceType::Boundary, 
                                         mfem::L2FaceValues::SingleValued);

  mfem::Vector x1(op->NumCols());
  mfem::Vector x2(op->NumCols());
  x1 = 5.0; 
  x2 = 0.0; 

  mfem::Vector y(op->NumRows());
  y = 5.0; 

  std::cout << "x1 before: "; x1.Print(std::cout);
  op->MultTranspose(y, x1);
  std::cout << "x1 after: "; x1.Print(std::cout);

  std::cout << "x2 before: "; x2.Print(std::cout);
  op->MultTranspose(y, x2);
  std::cout << "x2 after: "; x2.Print(std::cout);

  // output: 
  // x1 before: 5  5  5  5 5 5  5  5  5  5  5  5 5 5  5  5  5  5
  // x1 after: 15 10 10 10 5 5 15 10 10 10 10 15 5 5 10 10 10 15
  // x2 before: 0  0  0  0 0 0  0  0  0  0  0  0 0 0  0  0  0  0
  // x2 after: 10  5  5  5 0 0 10  5  5  5  5 10 0 0  5  5  5 10
}
```

Reading the documentation for `mfem::Operator`, `mfem::L2FaceRestriction`, one would expect `x1` and `x2` to be the same, but they are not. If this is intentional, it would be helpful to update the documentation accordingly. If not, `L2FaceRestriction::MultTranspose()` should be changed to be consistent with the semantics promised by `mfem::Operator`.

### PR DESCRIPTION
**Pull request by YohannDudouit**
Proposition to resolve #2328 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2329](https://github.com/mfem/mfem/pull/2329) | @YohannDudouit | @tzanio | @pazner + @artv3 | 06/12/21 | 06/17/21 | 06/25/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Comment by tzanio** [comment]
LGTM

File: fem/restriction.hpp
Comment (samuelpmishLLNL) [inline fem/restriction.hpp:123]: Could these descriptions be expanded to include the information needed to use this interface (e.g. what is expected of "x": number of entries, how they're packed, ... )?

I think `x` is expected to be an "L-vector" and `y` is a (face) "E-vector" in `Mult()` and vice versa in `AddMultTranspose()`, but I'm not sure.

A more complete high-level description could be included on the class definition too.

**Comment by tzanio** [comment]
The GitLab failure is bogus

File: fem/restriction.hpp
Comment (YohannDudouit) [reply fem/restriction.hpp]: I pushed some documentation, I added a lot of documentation to this in #2195 .

**Review by artv3 — APPROVED** [review]
Wonderful!

**Review by pazner — APPROVED** [review]
Nice contribution @YohannDudouit and thanks for pointing this out @samuelpmishLLNL.

**Comment by tzanio** [comment]
@samuelpmishLLNL I will wait on you approval before merging this in `next`

**Comment by tzanio** [comment]
Should we mention in `CHANGELOG`?

**Comment by artv3** [comment]
> Should we mention in `CHANGELOG`?

Sounds good to me!

**Comment by YohannDudouit** [comment]
> Should we mention in `CHANGELOG`?

We want to mention that the semantic of `MultTranspose` has changed to conform with `Operator` and that `AddMultTranspose` should be use instead?

**Comment by tzanio** [comment]
> @samuelpmishLLNL I will wait on you approval before merging this in `next`

ping: @samuelpmishLLNL

**Comment by samuelpmishLLNL** [comment]
Sorry for the delay. I'm not actually familiar with how to codevelop mfem in serac's build system to test this PR properly, but it appears to fix the discrepancy described in #2328, so it gets my approval.

**Review by samuelpmishLLNL — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by jamiebramwell** [comment]
Yeah, our TPL build solution makes co-development a bit tricky. I'm attempting to patch our build system to test it out, but I agree this should fix it. Thanks @YohannDudouit !

**Review by tzanio — APPROVED** [review]
Thanks for adding this @YohannDudouit.

@artv3, @pazner and @samuelpmishLLNL -- Thank you for the review.

**Pull request merged**
